src/tuner/nccl_ofi_model.c

```diff
@@ -0,0 +1,717 @@
+	 */
+	for (func = 0; func < NCCL_NUM_FUNCTIONS; func++) {
+		for (algo = 0; algo < NCCL_NUM_ALGORITHMS; algo++) {
+			for(proto = 0; proto < NCCL_NUM_PROTOCOLS; proto++) {
+				model_ctx->base_costs[func][algo][proto] =
+					nccl_ofi_tuner_compute_base_cost_p5_p5e(algo, proto);
+			}
+		}
+	}
+
+	return ncclSuccess;
+}
+
+
+/*****************************************************************************
+ *****************************************************************************
+ *         functions that are called by common tuner code start here
+ *****************************************************************************
+ *****************************************************************************/
+int is_model_supported(enum nccl_ofi_tuner_platform platform, size_t nRanks, size_t nNodes)
+{
+	if (platform == NCCL_OFI_TUNER_P5_P5E || platform == NCCL_OFI_TUNER_P5EN) {
+		return 1;
+	}
+
+	return 0;
+}
+
+ncclResult_t model_get_coll_info_internal_v3(nccl_ofi_tuner_context_t *ctx,
+					     ncclFunc_t collType,
+					     size_t nBytes,
+					     int numPipeOps,
+					     float **collCostTable,
+					     int numAlgo,
+					     int numProto,
+					     int *nChannels)
+{
+	nccl_ofi_tuner_model_context_t *model_ctx = (nccl_ofi_tuner_model_context_t *)ctx->type_ctx;
+
+	if (model_ctx == NULL) {
+		/* we do not update cost table. Fall back to NCCL's tuner */
+		NCCL_OFI_INFO(NCCL_TUNING, "Region Context is not ready. Fall back to NCCL's tuner.");
+		return ncclSuccess;
+	}
+
+	if (ctx->platform == NCCL_OFI_TUNER_P5_P5E) {
+		model_get_coll_info_internal_v3_p5_p5e(model_ctx, collType, nBytes, numPipeOps, collCostTable, nChannels);
+	} else if (ctx->platform == NCCL_OFI_TUNER_P5EN) {
+		model_get_coll_info_internal_v3_p5en(model_ctx, collType, nBytes, numPipeOps, collCostTable, nChannels);
+	} else {
+		/* Not valid platform. Do not update cost table. Fall back to NCCL's tuner */
+	}
+
+	return ncclSuccess;
+}
+
+ncclResult_t model_get_coll_info_internal_v2(nccl_ofi_tuner_context_t *ctx, ncclFunc_t collType, size_t nBytes,
+					     int collNetSupport, int nvlsSupport, int numPipeOps, int *algorithm,
+					     int *protocol, int *nChannels)
+{
+	nccl_ofi_tuner_model_context_t *model_ctx = (nccl_ofi_tuner_model_context_t *)ctx->type_ctx;
+
+	if (model_ctx == NULL) {
+		/* we do not update cost table. Fall back to NCCL's tuner */
+		NCCL_OFI_INFO(NCCL_TUNING, "Region Context is not ready. Fall back to NCCL's tuner.");
+		return ncclSuccess;
+	}
+
+	if (ctx->platform == NCCL_OFI_TUNER_P5_P5E) {
+		model_get_coll_info_internal_v2_p5_p5e(model_ctx, collType, nBytes, nvlsSupport,
+						       numPipeOps, algorithm, protocol, nChannels);
+	} else if (ctx->platform == NCCL_OFI_TUNER_P5EN) {
+		model_get_coll_info_internal_v2_p5en(model_ctx, collType, nBytes, nvlsSupport,
+						     numPipeOps, algorithm, protocol, nChannels);
+	} else {
+		/* Not valid platform. Do not update cost table. Fall back to NCCL's tuner */
```

Comment:
This should be done with platform_data values instead from a single function, so new platform additions will just need a new entry there for the tuner to work.

Reply:
Do you mean to have function pointers at `platform_data_map[]` in platform-aws.c? I am not sure whether adding function pointers is a good idea or not.

Or, do you mean to implement all the model logic inside a single function (ex. model_get_coll_info_internal_v2) so that platform specifics are just parameter values from `platform_data_map[]`? In this case, we will end up limiting flexibility in model logic. Not everything will be just decided by latency/bandwidth parameters. For example, we have a hard-coded manual selection for 4GB size for p5.

Reply:
> Or, do you mean to implement all the model logic inside a single function (ex. model_get_coll_info_internal_v2) so that platform specifics are just parameter values from platform_data_map[]? In this case, we will end up limiting flexibility in model logic. Not everything will be just decided by latency/bandwidth parameters. For example, we have a hard-coded manual selection for 4GB size for p5

Yeah, the latter. The inputs will always be the lowest-common-denominator of the model, IMO. If a platform needs a new input that just means the model has been updated and it is a no-op for older platforms. The 4gig override style on-offs can still be based on `nccl_net_ofi_get_product_name()`

Reply:
OK. I generally agree on the platform data based parameter setting, and have only a single `model_get_coll_info_internal_v3()` that has model logic, and p5/p5e vs. p5en different quirk can be if() based on platform type. For now, though, I suggest we add those platform specific parameter values inside Tuner code because adding `src/platform-aws.c` ends up pulling in many other files and symbol into `libnccl-ofi-tuner.so` too.